test/src/materials/SubdomainConstantMaterial.C

```diff
@@ -0,0 +1,57 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "SubdomainConstantMaterial.h"
+
+#include "MooseMesh.h"
+#include "FEProblem.h"
+
+template <>
+InputParameters
+validParams<SubdomainConstantMaterial>()
+{
+  InputParameters params = validParams<Material>();
+  params.addParam<MaterialPropertyName>("mat_prop_name", "Material property name");
+  params.addParam<std::vector<Real>>(
+      "values",
+      "Values of the material property on the subdomains (must be paired up with blocks)");
+  return params;
+}
+
+SubdomainConstantMaterial::SubdomainConstantMaterial(const InputParameters & parameters)
+  : Material(parameters),
+    _mat_prop(declareProperty<Real>(getParam<MaterialPropertyName>("mat_prop_name")))
+{
+  auto & blocks = getParam<std::vector<SubdomainName>>("block");
+  auto & values = getParam<std::vector<Real>>("values");
+  auto vec_ids = _mesh.getSubdomainIDs(blocks);
+  if (values.size() != vec_ids.size())
+    mooseError("number of values is not equal to the number of blocks");
+
+  for (unsigned int i = 0; i < vec_ids.size(); ++i)
+    _mapped_values[vec_ids[i]] = values[i];
+}
+
+void
+SubdomainConstantMaterial::subdomainSetup()
+{
+  // side material should not have subdomainSetup.
+  if (_bnd)
+    return;
+
+  _mat_prop.resize(_fe_problem.getMaxQps());
+  Real current_value = _mapped_values.at(_current_elem->subdomain_id());
+  for (_qp = 0; _qp < _fe_problem.getMaxQps(); ++_qp)
+    _mat_prop[_qp] = current_value;
+}
```

Comment:
@idaholab/moose-team - Looks like Yaqi has found a way to avoid calculating properties at each quadrature point. This is fairly advanced user code and we have to decide whether or not we should support this. I'm not completely against the idea but if we are going to do this, it probably should be done in MOOSE and supported as a first-class capability. It took me a minute to figure out why he was resizing the scalar property holder and looping to max qps... Anyway, this is what this code looks like.